nginx with pagespeed does not compile with gcc 5.3.0

Hello,

Since the gcc 5.3.0 update on Archlinux (see https://www.archlinux.org/news/c-abi-change/), I can't compile pagespeed. The problem occurs with version 1.9.32.10 and the new 1.9.32.11.

Here is the log : 
```bash
adding module in ../ngx_pagespeed
mod_pagespeed_dir=../ngx_pagespeed/psol/include
build_from_source=false
checking for psol ... not found
./configure: error: module ngx_pagespeed requires the pagespeed optimization library.
Look in objs/autoconf.err for more details.
```

And the objs/autoconf.err contains this : 
```
----------------------------------------
checking for psol

/tmp/ccThHr0f.o: In function `main':
autotest.cc:(.text.startup+0xfb): undefined reference to `net_instaweb::HtmlParse::StartParseId(base::BasicStringPiece<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > const&, base::BasicStringPiece<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > const&, net_instaweb::ContentType const&)'
collect2: error: ld returned 1 exit status
----------

#include <sys/types.h>
#include <unistd.h>

#include "pagespeed/kernel/base/string.h"
#include "pagespeed/kernel/base/string_writer.h"
#include "pagespeed/kernel/base/null_message_handler.h"
#include "pagespeed/kernel/html/html_parse.h"
#include "pagespeed/kernel/html/html_writer_filter.h"


int main() {

  GoogleString output_buffer;
  net_instaweb::StringWriter write_to_string(&output_buffer);

  net_instaweb::NullMessageHandler handler;
  net_instaweb::HtmlParse html_parse(&handler);
  net_instaweb::HtmlWriterFilter html_writer_filter(&html_parse);

  html_writer_filter.set_writer(&write_to_string);
  html_parse.AddFilter(&html_writer_filter);

  html_parse.StartParse("http:example.com");
  html_parse.ParseText(
      "<html ><body ><h1 >Test</h1 ><p>Test Text</p></body></html>\n");
  html_parse.FinishParse();

  printf("parsed as: %s", output_buffer.c_str());
    return 0;
}

----------
cc -march=native -O2 -pipe -fstack-protector-strong -D_GNU_SOURCE -D_FILE_OFFSET_BITS=64 -I ../ngx_pagespeed/psol/include -I ../ngx_pagespeed/psol/include/third_party/chromium/src -I ../ngx_pagespeed/psol/include/third_party/google-sparsehash/src -I ../ngx_pagespeed/psol/include/third_party/google-sparsehash/gen/arch/linux/x64/include -I ../ngx_pagespeed/psol/include/third_party/protobuf/src -I ../ngx_pagespeed/psol/include/third_party/re2/src -I ../ngx_pagespeed/psol/include/out/Release/obj/gen -I ../ngx_pagespeed/psol/include/out/Release/obj/gen/protoc_out/instaweb -I ../ngx_pagespeed/psol/include/third_party/apr/src/include -I ../ngx_pagespeed/psol/include/third_party/aprutil/src/include -I ../ngx_pagespeed/psol/include/third_party/apr/gen/arch/linux/x64/include -I ../ngx_pagespeed/psol/include/third_party/aprutil/gen/arch/linux/x64/include -o objs/autotest objs/autotest.cc -lstdc++ ../ngx_pagespeed/psol/lib/Release/linux/x64/pagespeed_automatic.a -lrt -pthread -lm
----------
```

If you want the complete file : http://pastebin.com/QpF4UfnR

Unfortunately, the commit https://github.com/pagespeed/ngx_pagespeed/pull/1048/files did not work for this problem.

autoconf.err end : 
```
----------------------------------------
checking for psol

/tmp/ccuXzaYW.o: In function `main':
autotest.cc:(.text.startup+0xfb): undefined reference to `net_instaweb::HtmlParse::StartParseId(base::BasicStringPiece<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > const&, base::BasicStringPiece<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > const&, net_instaweb::ContentType const&)'
collect2: error: ld returned 1 exit status
----------

#include <sys/types.h>
#include <unistd.h>

#include "pagespeed/kernel/base/string.h"
#include "pagespeed/kernel/base/string_writer.h"
#include "pagespeed/kernel/base/null_message_handler.h"
#include "pagespeed/kernel/html/html_parse.h"
#include "pagespeed/kernel/html/html_writer_filter.h"


int main() {

  GoogleString output_buffer;
  net_instaweb::StringWriter write_to_string(&output_buffer);

  net_instaweb::NullMessageHandler handler;
  net_instaweb::HtmlParse html_parse(&handler);
  net_instaweb::HtmlWriterFilter html_writer_filter(&html_parse);

  html_writer_filter.set_writer(&write_to_string);
  html_parse.AddFilter(&html_writer_filter);

  html_parse.StartParse("http:example.com");
  html_parse.ParseText(
      "<html ><body ><h1 >Test</h1 ><p>Test Text</p></body></html>\n");
  html_parse.FinishParse();

  printf("parsed as: %s", output_buffer.c_str());
    return 0;
}

----------
cc -march=native -O2 -pipe -fstack-protector-strong -D_GNU_SOURCE -D_FILE_OFFSET_BITS=64 -I ../ngx_pagespeed/psol/include -I ../ngx_pagespeed/psol/include/third_party/chromium/src -I ../ngx_pagespeed/psol/include/third_party/google-sparsehash/src -I ../ngx_pagespeed/psol/include/third_party/google-sparsehash/gen/arch/linux/x64/include -I ../ngx_pagespeed/psol/include/third_party/protobuf/src -I ../ngx_pagespeed/psol/include/third_party/re2/src -I ../ngx_pagespeed/psol/include/out/Release/obj/gen -I ../ngx_pagespeed/psol/include/out/Release/obj/gen/protoc_out/instaweb -I ../ngx_pagespeed/psol/include/third_party/apr/src/include -I ../ngx_pagespeed/psol/include/third_party/aprutil/src/include -I ../ngx_pagespeed/psol/include/third_party/apr/gen/arch/linux/x64/include -I ../ngx_pagespeed/psol/include/third_party/aprutil/gen/arch/linux/x64/include -o objs/autotest objs/autotest.cc -lstdc++ ../ngx_pagespeed/psol/lib/Release/linux/x64/pagespeed_automatic.a -lrt -pthread -lm
----------
```

I see, looks like it didn't actually pass it in to the configure tests.
That's what I get for only testing this with a compiler that doesn't need
it.
I think I commented on config tests not getting the same flags somewhere
else, looking it up.


On Thu, Dec 10, 2015 at 3:10 PM, rfnx <notifications@github.com> wrote:

> Unfortunately, the commit
> https://github.com/pagespeed/ngx_pagespeed/pull/1048/files did not work
> for this problem.
>
> autoconf.err end :
>
> ----------------------------------------
> checking for psol
>
> /tmp/ccuXzaYW.o: In function `main':
> autotest.cc:(.text.startup+0xfb): undefined reference to `net_instaweb::HtmlParse::StartParseId(base::BasicStringPiece<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > const&, base::BasicStringPiece<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > const&, net_instaweb::ContentType const&)'
> collect2: error: ld returned 1 exit status
> ----------
>
> #include <sys/types.h>
> #include <unistd.h>
>
> #include "pagespeed/kernel/base/string.h"
> #include "pagespeed/kernel/base/string_writer.h"
> #include "pagespeed/kernel/base/null_message_handler.h"
> #include "pagespeed/kernel/html/html_parse.h"
> #include "pagespeed/kernel/html/html_writer_filter.h"
>
>
> int main() {
>
>   GoogleString output_buffer;
>   net_instaweb::StringWriter write_to_string(&output_buffer);
>
>   net_instaweb::NullMessageHandler handler;
>   net_instaweb::HtmlParse html_parse(&handler);
>   net_instaweb::HtmlWriterFilter html_writer_filter(&html_parse);
>
>   html_writer_filter.set_writer(&write_to_string);
>   html_parse.AddFilter(&html_writer_filter);
>
>   html_parse.StartParse("http:example.com");
>   html_parse.ParseText(
>       "<html ><body ><h1 >Test</h1 ><p>Test Text</p></body></html>\n");
>   html_parse.FinishParse();
>
>   printf("parsed as: %s", output_buffer.c_str());
>     return 0;
> }
>
> ----------
> cc -march=native -O2 -pipe -fstack-protector-strong -D_GNU_SOURCE -D_FILE_OFFSET_BITS=64 -I ../ngx_pagespeed/psol/include -I ../ngx_pagespeed/psol/include/third_party/chromium/src -I ../ngx_pagespeed/psol/include/third_party/google-sparsehash/src -I ../ngx_pagespeed/psol/include/third_party/google-sparsehash/gen/arch/linux/x64/include -I ../ngx_pagespeed/psol/include/third_party/protobuf/src -I ../ngx_pagespeed/psol/include/third_party/re2/src -I ../ngx_pagespeed/psol/include/out/Release/obj/gen -I ../ngx_pagespeed/psol/include/out/Release/obj/gen/protoc_out/instaweb -I ../ngx_pagespeed/psol/include/third_party/apr/src/include -I ../ngx_pagespeed/psol/include/third_party/aprutil/src/include -I ../ngx_pagespeed/psol/include/third_party/apr/gen/arch/linux/x64/include -I ../ngx_pagespeed/psol/include/third_party/aprutil/gen/arch/linux/x64/include -o objs/autotest objs/autotest.cc -lstdc++ ../ngx_pagespeed/psol/lib/Release/linux/x64/pagespeed_automatic.a -lrt -pthread -lm
> ----------
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pagespeed/ngx_pagespeed/issues/1066#issuecomment-163736413>
> .
>

Could you please also try adding a line saying
CC_TEST_FLAGS=="$CC_TEST_FLAGS -D_GLIBCXX_USE_CXX11_ABI=0"
in ngx_pagespeed/config file at line 86, right where the patch added new stuff?


On Thu, Dec 10, 2015 at 3:18 PM, Maksim Orlovich <morlovich@google.com> wrote:
> I see, looks like it didn't actually pass it in to the configure tests.
> That's what I get for only testing this with a compiler that doesn't need
> it.
> I think I commented on config tests not getting the same flags somewhere
> else, looking it up.
>
>
> On Thu, Dec 10, 2015 at 3:10 PM, rfnx <notifications@github.com> wrote:
>>
>> Unfortunately, the commit
>> https://github.com/pagespeed/ngx_pagespeed/pull/1048/files did not work for
>> this problem.
>>
>> autoconf.err end :
>>
>> ----------------------------------------
>> checking for psol
>>
>> /tmp/ccuXzaYW.o: In function `main':
>> autotest.cc:(.text.startup+0xfb): undefined reference to
>> `net_instaweb::HtmlParse::StartParseId(base::BasicStringPiece<std::__cxx11::basic_string<char,
>> std::char_traits<char>, std::allocator<char> > > const&,
>> base::BasicStringPiece<std::__cxx11::basic_string<char,
>> std::char_traits<char>, std::allocator<char> > > const&,
>> net_instaweb::ContentType const&)'
>> collect2: error: ld returned 1 exit status
>> ----------
>>
>> #include <sys/types.h>
>> #include <unistd.h>
>>
>> #include "pagespeed/kernel/base/string.h"
>> #include "pagespeed/kernel/base/string_writer.h"
>> #include "pagespeed/kernel/base/null_message_handler.h"
>> #include "pagespeed/kernel/html/html_parse.h"
>> #include "pagespeed/kernel/html/html_writer_filter.h"
>>
>>
>> int main() {
>>
>>   GoogleString output_buffer;
>>   net_instaweb::StringWriter write_to_string(&output_buffer);
>>
>>   net_instaweb::NullMessageHandler handler;
>>   net_instaweb::HtmlParse html_parse(&handler);
>>   net_instaweb::HtmlWriterFilter html_writer_filter(&html_parse);
>>
>>   html_writer_filter.set_writer(&write_to_string);
>>   html_parse.AddFilter(&html_writer_filter);
>>
>>   html_parse.StartParse("http:example.com");
>>   html_parse.ParseText(
>>       "<html ><body ><h1 >Test</h1 ><p>Test Text</p></body></html>\n");
>>   html_parse.FinishParse();
>>
>>   printf("parsed as: %s", output_buffer.c_str());
>>     return 0;
>> }
>>
>> ----------
>> cc -march=native -O2 -pipe -fstack-protector-strong -D_GNU_SOURCE
>> -D_FILE_OFFSET_BITS=64 -I ../ngx_pagespeed/psol/include -I
>> ../ngx_pagespeed/psol/include/third_party/chromium/src -I
>> ../ngx_pagespeed/psol/include/third_party/google-sparsehash/src -I
>> ../ngx_pagespeed/psol/include/third_party/google-sparsehash/gen/arch/linux/x64/include
>> -I ../ngx_pagespeed/psol/include/third_party/protobuf/src -I
>> ../ngx_pagespeed/psol/include/third_party/re2/src -I
>> ../ngx_pagespeed/psol/include/out/Release/obj/gen -I
>> ../ngx_pagespeed/psol/include/out/Release/obj/gen/protoc_out/instaweb -I
>> ../ngx_pagespeed/psol/include/third_party/apr/src/include -I
>> ../ngx_pagespeed/psol/include/third_party/aprutil/src/include -I
>> ../ngx_pagespeed/psol/include/third_party/apr/gen/arch/linux/x64/include -I
>> ../ngx_pagespeed/psol/include/third_party/aprutil/gen/arch/linux/x64/include
>> -o objs/autotest objs/autotest.cc -lstdc++
>> ../ngx_pagespeed/psol/lib/Release/linux/x64/pagespeed_automatic.a -lrt
>> -pthread -lm
>> ----------
>>
>> —
>> Reply to this email directly or view it on GitHub.
>
>

--- a/config
+++ b/config
@@ -98,6 +98,7 @@ CFLAGS="$CFLAGS $FLAG_MARCH"
 # to our libpsol.a
 # See https://gcc.gnu.org/onlinedocs/libstdc++/manual/using_dual_abi.html
 CFLAGS="$CFLAGS -D_GLIBCXX_USE_CXX11_ABI=0"
+CC_TEST_FLAGS="$CC_TEST_FLAGS -D_GLIBCXX_USE_CXX11_ABI=0"
 
 case "$NGX_GCC_VER" in
   4.8*)

